File: core-java/src/main/java/com/kakao/actionbase/core/java/metadata/v3/AliasDescriptor.java
package com.kakao.actionbase.core.java.metadata.v3;

import com.kakao.actionbase.core.java.edge.EdgePayload;
import com.kakao.actionbase.core.java.edge.EdgeState;
import com.kakao.actionbase.core.java.metadata.AliasId;
import com.kakao.actionbase.core.java.metadata.ImmutableAliasId;
import com.kakao.actionbase.core.java.metadata.ImmutableTableId;
import com.kakao.actionbase.core.java.metadata.TableId;

import org.immutables.value.Value;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

@Value.Immutable
@JsonSerialize(as = ImmutableAliasDescriptor.class)
@JsonDeserialize(as = ImmutableAliasDescriptor.class)
public interface AliasDescriptor extends Descriptor<AliasId> {

  String database();

  String alias();

  String table();

  @Override
  @JsonIgnore
  @Value.Derived
  @Value.Auxiliary
  default AliasId id() {
    return ImmutableAliasId.of(tenant(), database(), alias());
  }

  @JsonIgnore
  @Value.Derived
  @Value.Auxiliary
  default TableId tableId() {
    return ImmutableTableId.of(tenant(), database(), table());
  }

  @Override
  @JsonIgnore
  @Value.Auxiliary
  default EdgePayload toEdgePayload() {
    return Aliases.toEdge(this);
  }

  @Override
  @JsonIgnore
  @Value.Auxiliary
  default EdgeState toEdgeState() {
    return null;
  }

  @JsonIgnore
  @Value.Auxiliary
  default AliasPayload toPayload() {
    return ImmutableAliasPayload.builder().alias(alias()).table(table()).comment(comment()).build();
  }
}

File: core/src/main/kotlin/com/kakao/actionbase/core/metadata/AliasDescriptor.kt
package com.kakao.actionbase.core.metadata

import com.kakao.actionbase.core.Constants

import com.fasterxml.jackson.annotation.JsonIgnore

data class AliasDescriptor(
    override val tenant: String,
    val database: String,
    val alias: String,
    val table: String,
    override val active: Boolean = true,
    override val comment: String = Constants.DEFAULT_COMMENT,
    override val revision: Long = Constants.DEFAULT_REVISION,
    override val createdAt: Long = Constants.DEFAULT_CREATED_AT,
    override val createdBy: String = Constants.DEFAULT_CREATED_BY,
    override val updatedAt: Long = Constants.DEFAULT_UPDATED_AT,
    override val updatedBy: String = Constants.DEFAULT_UPDATED_BY,
) : V3Descriptor<AliasId> {
    @JsonIgnore
    override val id: AliasId = AliasId(tenant, database, alias)
}
